api/turing/batch/runner/ensembling.go

```diff
@@ -0,0 +1,322 @@
+package batchrunner
+
+import (
+	"time"
+
+	mlp "github.com/gojek/mlp/client"
+	batchcontroller "github.com/gojek/turing/api/turing/batch/controller"
+	"github.com/gojek/turing/api/turing/imagebuilder"
+	"github.com/gojek/turing/api/turing/internal/testutils"
+	"github.com/gojek/turing/api/turing/log"
+	"github.com/gojek/turing/api/turing/models"
+	"github.com/gojek/turing/api/turing/service"
+)
+
+type ensemblingJobRunner struct {
+	ensemblingController      batchcontroller.EnsemblingController
+	ensemblingJobService      service.EnsemblingJobService
+	mlpService                service.MLPService
+	imageBuilder              imagebuilder.ImageBuilder
+	injectGojekLabels         bool
+	environment               string
+	batchSize                 int
+	maxRetryCount             int
+	imageBuildTimeoutDuration time.Duration
+}
+
+var (
+	labelTeamName         = "gojek.com/team"
+	labelStreamName       = "gojek.com/stream"
+	labelAppName          = "gojek.com/app"
+	labelEnvironment      = "gojek.com/environment"
+	labelOrchestratorName = "gojek.com/orchestrator"
+	valueOrchestratorName = "turing"
+)
+
+// NewBatchEnsemblingJobRunner creates a new batch ensembling job
+// This service controls the orchestration of batch ensembling jobs.
+func NewBatchEnsemblingJobRunner(
+	ensemblingController batchcontroller.EnsemblingController,
+	ensemblingJobService service.EnsemblingJobService,
+	mlpService service.MLPService,
+	imageBuilder imagebuilder.ImageBuilder,
+	injectGojekLabels bool,
+	environment string,
+	batchSize int,
+	maxRetryCount int,
+	imageBuildTimeoutDuration time.Duration,
+) BatchJobRunner {
+	return &ensemblingJobRunner{
+		ensemblingController:      ensemblingController,
+		ensemblingJobService:      ensemblingJobService,
+		mlpService:                mlpService,
+		imageBuilder:              imageBuilder,
+		injectGojekLabels:         injectGojekLabels,
+		environment:               environment,
+		batchSize:                 batchSize,
+		maxRetryCount:             maxRetryCount,
+		imageBuildTimeoutDuration: imageBuildTimeoutDuration,
+	}
+}
+
+func (r *ensemblingJobRunner) Run() {
+	r.processJobs()
+	r.updateStatus()
+}
+
+func (r *ensemblingJobRunner) updateStatus() {
+	// Here we want to check all non terminal cases.
+	// i.e. JobBuildingImage but with updated_at timeout
+	//		JobRunning
+	optionsJobRunning := service.EnsemblingJobListOptions{
+		PaginationOptions: service.PaginationOptions{
+			Page:     testutils.NullableInt(1),
+			PageSize: &r.batchSize,
+		},
+		Statuses: []models.Status{
+			models.JobRunning,
+		},
+	}
+	r.processEnsemblingJobs(optionsJobRunning)
+
+	// To ensure that we only pickup jobs that weren't picked up, we check the last updated at.
+	updatedAtBefore := time.Now().Add(r.imageBuildTimeoutDuration * -1)
```

Comment:
Well, for one updated_at < now is a tautology 😄.

For the weird cryptic explanation:

Case 1: Job is actually being watched by a live goroutine.
```
             <---imageBuildTimeoutDuration---> 
------------|-----------------|---------------|---------> time
          job start          now       timeout cut off
          (updated at)
```
We don't want to pick up this situation since the image building might take a long time.

Case 2: job hanged or Turing API unexpectedly terminated.
```
             <---imageBuildTimeoutDuration---> 
------------|---------------------------------|------> time
          job start                          now
          (updated at)
```
Here turing API might have crashed and we probably want to retry this situation.

Reply:
Haha. Thanks for the clarification. Can this be captured in comments in the code?

Reply:
ok sure.